pyro/infer/trace_mmd.py

```diff
@@ -0,0 +1,169 @@
+                 max_iarange_nesting=None,  # DEPRECATED
+                 vectorize_particles=False,
+                 strict_enumeration_warning=True,
+                 ignore_jit_warnings=False,
+                 retain_graph=None):
+        super().__init__(
+            num_particles, max_plate_nesting, max_iarange_nesting, vectorize_particles,
+            strict_enumeration_warning, ignore_jit_warnings, retain_graph,
+        )
+        self._kernel = None
+        self._mmd_scale = None
+        self.kernel = kernel
+        self.mmd_scale = mmd_scale
+
+    @property
+    def kernel(self):
+        return self._kernel
+
+    @kernel.setter
+    def kernel(self, kernel):
+        if isinstance(kernel, dict):
+            self._kernel = kernel
+        elif isinstance(kernel, pyro.contrib.gp.kernels.kernel.Kernel):
+            self._kernel = defaultdict(lambda: kernel)
+        else:
+            raise TypeError("`kernel` should be an instance of `pyro.contrib.gp.kernels.kernel.Kernel`")
+
+    @property
+    def mmd_scale(self):
+        return self._mmd_scale
+
+    @mmd_scale.setter
+    def mmd_scale(self, mmd_scale):
+        if isinstance(mmd_scale, dict):
+            self._mmd_scale = mmd_scale
+        elif isinstance(mmd_scale, (int, float)):
+            self._mmd_scale = defaultdict(lambda: float(mmd_scale))
+        else:
+            raise TypeError("`mmd_scale` should be either float, or a dict of floats")
+
+    def _differentiable_loss_parts(self, model, guide, *args, **kwargs):
+        loglikelihood = 0.0
+        penalty = 0.0
+        model_trace_independent = poutine.trace(model, graph_type="flat").get_trace(*args, **kwargs)
+        for model_trace, guide_trace in self._get_traces(model, guide, *args, **kwargs):
+            loglikelihood_particle = 0.0
+            penalty_particle = 0.0
+            for name, model_site in model_trace.nodes.items():
+                if model_site['type'] == 'sample':
+                    if name in guide_trace:
+                        guide_site = guide_trace.nodes[name]
+                        independent_model_site = model_trace_independent.nodes[name]
+                        model_samples = independent_model_site['value']
+                        guide_samples = guide_site['value']
+                        divergence = _compute_mmd(
+                            model_samples.view(model_samples.size(0), -1),
+                            guide_samples.view(guide_samples.size(0), -1),
+                            kernel=self._kernel[name]
+                        )
+                        penalty_particle = penalty_particle + self._mmd_scale[name] * divergence
+                    else:
+                        loglikelihood_particle = loglikelihood_particle + model_site['log_prob_sum']
+            loglikelihood = loglikelihood_particle / self.num_particles + loglikelihood
+            penalty = penalty_particle / self.num_particles + penalty
```

Comment:
Since `_compute_mmd` uses `.mean()` to reduce over the particle dimension, I believe this is including an extra factor of `1 / self.num_particles` in `penalty`.

Reply:
I didn't know about particle dimension. It appears only if `vectorize_particles=True`. As a result, there was a severe bug in MMD computation. Thanks for a tip!